Use unauthenticated root path

paulomarg opened the pull request:
### WHY are these changes introduced?

The current implementation of the app still relied on 3rd party cookies for the initial `/` load to fetch the app skeleton, which still caused it to fail if one attempted to reach the app by clicking the Admin link to it.

### WHAT is this pull request doing?

I added an sqlite database to the app so we can store the shops to which it's been installed. Since we now have a means of persisting data, I also went ahead and added a `CustomSessionStorage` object which also leverages it to provide persistent sessions.

paulomarg commented on server/server.js:
```diff
@@ -9,6 +9,11 @@ import Shopify, { ApiVersion } from "@shopify/shopify-api";
+import { ShopModel } from "./db/models/shop";
+import { SessionModel } from "./db/models/session";
+
+import "./db/seed";
```
This should definitely be a `yarn` script, I just wanted to get thoughts on using a db at all before looking into doing that!

paulomarg commented on server/db/seed.js:
```diff
@@ -0,0 +1,13 @@
+  await Db.query("CREATE TABLE IF NOT EXISTS shops (shop TEXT)");
+  await Db.query("CREATE UNIQUE INDEX shops_unique_shop_index ON shops(shop)");
+
+  await Db.query("CREATE TABLE IF NOT EXISTS sessions (id TEXT, data JSON)");
```
Also, I wonder if we want to have the `Session` object provide a serialization / deserialization option or if JSON parse / stringify is enough.

thecodepixi replied:
I think the `JSON.parse` is fine in this case. Maybe in a real-world app someone might want to write a serializer but I don't think it's totally necessary here. `JSON.stringify` to save and `JSON.parse` to load get the job done.

mkevinosullivan reviewed (approved):
Since it's our main example application, it's probably not a bad idea to demonstrate that session storage (and storage of the shop names on which the app is installed) is necessary for a robust app.

The alternative view, that it's just an example and we should keep it simple, would dictate that the db stuff isn't needed ... but I'm leaning towards keeping it in.

thecodepixi reviewed (approved):
I essentially echo Kevin's sentiment here, and I think having a working example of a `CustomSessionStorage` solution is valuable. LGTM! 👍

paulomarg commented:
Cool, I'll turn the seed import into a yarn script and add these changes to the tutorial as well! Thanks everyone.

tanema commented:
Should we be making these types of code commitments when we could fix this issue with documentation? This feels like a section of code that will most likely be an issue with how to change this over to a different storage method. This seems like a lot of code for a single type of app access

paulomarg commented:
> Should we be making these types of code commitments when we could fix this issue with documentation? This feels like a section of code that will most likely be an issue with how to change this over to a different storage method. This seems like a lot of code for a single type of app access

That was something I pondered myself, but we do need to store which stores have installed the app somehow, otherwise we won't be able to know when to perform OAuth. The alternative is:
- Go back to `MemorySessionStorage` to get rid of `SessionModel` (I just added a db-based one to showcase how it works)
- Replace the `ShopModel` actions with a server-side hash which holds the apps in memory. It works for development but it costs memory and it gets destroyed when the server is restarted - which would force all apps to go through OAuth, not necessarily a big deal.

I added a comment on `MemorySessionStorage` to the tutorial, so I could simply add another one indicating that devs might want to consider storing this value somewhere. Do you think we can convey that clearly with documentation only?

mkevinosullivan reviewed (approved):
LGTM! 🚀
